src/Http/Routing/src/RouteGroupContext.cs

```diff
@@ -11,39 +11,35 @@ namespace Microsoft.AspNetCore.Routing;
 /// </summary>
 public sealed class RouteGroupContext
 {
-    /// <summary>
-    /// Constructs a new <see cref="RouteGroupContext"/> instance.
-    /// </summary>
-    /// <param name="prefix">The full group prefix. See <see cref="Prefix"/>.</param>
-    /// <param name="conventions">All conventions added to a parent group. See <see cref="Conventions"/>.</param>
-    /// <param name="applicationServices">Application services. See <see cref="ApplicationServices"/>.</param>
-    public RouteGroupContext(RoutePattern prefix, IReadOnlyList<Action<EndpointBuilder>> conventions, IServiceProvider applicationServices)
-    {
-        ArgumentNullException.ThrowIfNull(prefix);
-        ArgumentNullException.ThrowIfNull(conventions);
-        ArgumentNullException.ThrowIfNull(applicationServices);
-
-        Prefix = prefix;
-        Conventions = conventions;
-        ApplicationServices = applicationServices;
-    }
-
     /// <summary>
     /// Gets the <see cref="RouteEndpoint.RoutePattern"/> which should prefix the <see cref="RouteEndpoint.RoutePattern"/> of all <see cref="RouteEndpoint"/> instances
     /// returned by the call to <see cref="EndpointDataSource.GetGroupedEndpoints(RouteGroupContext)"/>. This accounts for nested groups and gives the full group prefix
     /// not just the prefix supplied to the innermost call to <see cref="EndpointRouteBuilderExtensions.MapGroup(IEndpointRouteBuilder, RoutePattern)"/>.
     /// </summary>
-    public RoutePattern Prefix { get; }
+    public required RoutePattern Prefix { get; init; }
 
     /// <summary>
     /// Gets all conventions added to ancestor <see cref="RouteGroupBuilder"/> instances returned from <see cref="EndpointRouteBuilderExtensions.MapGroup(IEndpointRouteBuilder, RoutePattern)"/>
     /// via <see cref="IEndpointConventionBuilder.Add(Action{EndpointBuilder})"/>. These should be applied in order when building every <see cref="RouteEndpoint"/>
     /// returned from <see cref="EndpointDataSource.GetGroupedEndpoints(RouteGroupContext)"/>.
     /// </summary>
-    public IReadOnlyList<Action<EndpointBuilder>> Conventions { get; }
+    public IReadOnlyList<Action<EndpointBuilder>> Conventions { get; init; } = Array.Empty<Action<EndpointBuilder>>();
+
+    /// <summary>
+    /// Gets all conventions added to ancestor <see cref="RouteGroupBuilder"/> instances returned from <see cref="EndpointRouteBuilderExtensions.MapGroup(IEndpointRouteBuilder, RoutePattern)"/>
+    /// via <see cref="IEndpointConventionBuilder.Add(Action{EndpointBuilder})"/>. These should be applied in LIFO order when building every <see cref="RouteEndpoint"/>
+    /// returned from <see cref="EndpointDataSource.GetGroupedEndpoints(RouteGroupContext)"/>.
+    /// </summary>
+    public IReadOnlyList<Action<EndpointBuilder>> FinallyConvnentions { get; init; } = Array.Empty<Action<EndpointBuilder>>();
```

Comment:
```suggestion
    public IReadOnlyList<Action<EndpointBuilder>> FinallyConventions { get; init; } = Array.Empty<Action<EndpointBuilder>>();
```

Needs to be fixed everywhere 😄

Reply:
*roblox death sound*

I'll take the L on this...